### PR TITLE
add allow_cors decorator and allow CORS for new case API

### PR DESCRIPTION
**Pull request by czue**
## Summary

This enables CORS headers on the new case API via a new decorator function.

The standard answer to Django + CORS seems to be [django-cors-headers](https://pypi.org/project/django-cors-headers/) but it seems a rather blunt instrument that isn't designed to make things configurable at the view level, more at the project level. So I went with a much simpler approach inspired by [our current TastyPie implementation](https://github.com/dimagi/commcare-hq/pull/4564)

Very open to feedback on this approach though - this isn't an area of expertise of mine.

## Feature Flag

Currently only affects the v0.6 case API

## Product Description

External applications can integrate with the v0.6 case API.

## Safety Assurance

- [x] Risk label is set correctly
- [x] All migrations are backwards compatible and won't block deploy
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change
- [x] If QA is part of the safety story, the "Awaiting QA" label is used
- [x] I have confidence that this PR will not introduce a regression for the reasons below

New code is well tested and touches to old code were very rote refactor/cleanups.

### Automated test coverage

Added tests for new functionality. Existing APIs also have tests.

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### CODE REVIEW

**Comment by czue** [comment]
I'm also happy to try and add tests for the decorator and reconcile some of this code with the tastypie implementation if either of those feel like worthwhile work. Mostly I was interested in getting a quick gut-check on this approach before investing too much work into it.

File: corehq/apps/domain/decorators.py [hunk 1]
```diff
@@ -675,3 +675,21 @@ def _wrapped(*args, **kwargs):
...
 
     return _dec
+
+
+def allow_cors(allowed_methods):
+    def decorator(view_func):
+        @wraps(view_func)
+        def wrapped_view(request, *args, **kwargs):
+            if request.method == "OPTIONS":
+                response = HttpResponse(allowed_methods)
```
Comment (millerdev) [inline corehq/apps/domain/decorators.py:685]: Should this be `JsonResponse`? Didn't know it was valid to pass a list as response `content`.

**Review by millerdev — COMMENTED** [review]
:+1: for tests

File: corehq/apps/domain/decorators.py [hunk 2]
```diff
@@ -675,3 +675,21 @@ def _wrapped(*args, **kwargs):
         return _wrapped
 
     return _dec
+
+
+def allow_cors(allowed_methods):
```
Comment (esoergel) [inline corehq/apps/domain/decorators.py:680]: I see the tastypie implementation also verifies that the request method is one of the allowed methods - do you think this should do the same, or does that fall to things like `require_POST` or as in the case api, the body of the view.

**Review by esoergel — APPROVED** [review]
I like this approach - I also looked in to `django_cors_headers` yesterday and thought it looked like a bad fit for our use-case.  +1 to unifying with the `CorsResourceMixin` if straightforward, though it uses a short-circuit logic that might not be easily amenable to this approach.

File: corehq/apps/domain/decorators.py (hunk 2 above)
Comment (czue) [reply corehq/apps/domain/decorators.py]: good call. added this

File: corehq/apps/domain/decorators.py (hunk 1 above)
Comment (czue) [reply corehq/apps/domain/decorators.py]: This does work, though wasn't intentional ([here's why if you're curious](https://github.com/django/django/blob/main/django/http/response.py#L353-L357) - maybe related to chunked encoding). Anyway, I read up and the response body of an `OPTIONS` request is undefined, so I just removed it.

The API version returned a comma separated list of the methods, which I left untouched.

```
$ curl -X OPTIONS https://www.commcarehq.org/a/sabbatical/api/v0.5/user/ -i
HTTP/1.1 200 OK
Date: Tue, 06 Apr 2021 08:37:31 GMT
Content-Type: text/html; charset=utf-8
Content-Length: 8
Connection: keep-alive
Server: nginx
Access-Control-Allow-Origin: *
Access-Control-Allow-Headers: Content-Type, Authorization
Allow: GET,POST
Strict-Transport-Security: max-age=300; includeSubDomains
X-Content-Type-Options: nosniff
X-XSS-Protection: 1; mode=block
HTTP_X_OPENROSA_VERSION: 1.0
X-Frame-Options: DENY
Vary: Accept-Language, Cookie
Content-Language: en
Cache-Control: private, no-cache, no-store, must-revalidate, proxy-revalidate
Expires: Thu, 01 Dec 1994 16:00:00 GMT
Pragma: no-cache
Age=1209600; Path=/; SameSite=Lax; Secure
Strict-Transport-Security: max-age=300
X-XSS-Protection: 1; mode=block
X-Content-Type-Options: nosniff

GET,POST
```

**Comment by czue** [comment]
@millerdev @esoergel finally got around to coming back to this. requesting your re-review

**Review by millerdev — APPROVED** [review]

File: corehq/apps/api/resources/__init__.py [hunk 3]
```diff
@@ -64,10 +65,7 @@ class CorsResourceMixin(object):
 
     def create_response(self, *args, **kwargs):
         response = super(CorsResourceMixin, self).create_response(*args, **kwargs)
-        response['Access-Control-Allow-Origin'] = '*'
-        response['Access-Control-Allow-Headers'] = 'Content-Type, Authorization'
-
-        return response
+        return add_cors_headers_to_response(response)
```
Comment (esoergel) [inline corehq/apps/api/resources/__init__.py:68]: :star:

**Review by esoergel — COMMENTED** [review]
:+1:

**Pull request merged**